### PR TITLE
Ilya - Tools table

### PR DESCRIPTION
**Pull request by ilyaflaks**
tool-list-task

# Description
Adds functionality for the user to navigate to [/bmdashboard/tools](http://localhost:3000/bmdashboard/tools), and view a list of all tool items currently in the system. The table can be sorted by tool name or by project and the results can be filtered by tool name. Update History and Purchase History can be viewed for each item in a modal.

## Related PRS (if any):
This backend PR is related to the [#2193](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2193) frontend PR.
To test this backend PR you need to checkout the #2193 frontend PR, branch **ilya_tools_table_fe**

## Main changes explained:
- largeItemBaseSchema changed to have a "condition" field
- buildingTool schema changed to have specific values for the purchaseStatus field
- fetchAllTools function added to bmToolController to query the database and return a full list of tool items and populate certain fields
- A new route added to call fetchAllTools when a GET request is made to /api/tools

## How to test:
1. Check out the current branch. Make sure you are using Node version 14.17.6. Run **npm install** and **npm run build && npm start** to run the backend locally
2. Check out **ilya_tools_table_fe** branch on the front end. Run commands **npm install** and **npm run start:local** to launch the front end application
3. On the front end log in as any user
4. Go to [/bmdashboard](http://localhost:3000/bmdashboard) and log in again
5. Navigate to [/bmdashboard/tools](http://localhost:3000/bmdashboard/tools). A list of all tools currently in the system should display.
6. Test the filtering by Project and Tool function (the dropdowns above the table)
7. Test sorting by project and by name (the arrows next to "Project" and "Name" in the table header
8. Pressing the View button under Updates should open an Updates Record modal (not all items have information to display)
9. Pressing the View button under Purchases should open an Purchase Record modal (not all items have information to display)
10. Pressing the pencil should open an Update Tool Form modal. The content of this modal is outside of the scope of this task

## Screenshots or video

https://github.com/OneCommunityGlobal/HGNRest/assets/105448137/891fe190-e5c4-492d-86c0-c7362769957a

### CODE REVIEW

**Review by mSharifHub — CHANGES_REQUESTED** [review]
The controllers needs to be refractor for clarity and mostly dont use try and catch along with promise chain. You can use try and catch for better clarity and  provide more information on the error handling.

**Review by alla0810 — APPROVED** [review]
Left PR review [on the frontend side](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2193#pullrequestreview-2035737972).
Great work!
Everything works great!

File: src/controllers/bmdashboard/bmToolController.js [hunk 1]
```diff
@@ -1,6 +1,53 @@
...
+                select: '_id firstName lastName',
+            },
+            {
+                path: 'responsibleUser',
+                select: '_id firstName lastName',
+        }],
+        },
+        ])
+        .exec()
+        .then(results => res.status(200).send(results))
```
[the remark this answers is not in the document]
Comment (ilyaflaks) [reply src/controllers/bmdashboard/bmToolController.js]: @mSharifHub  Good observation about try/catch. I have removed it in the latest commit and added a more explicit error message. The format for the populate function is a matter of preference, but I moved it outside of the query anyway.
Thank you for your review

**Review by Ramaruva — APPROVED** [review]
Reviewed changes in frontend related to this branch and changes are working as expected

**Review by Parth-tech — APPROVED** [review]
Hey Llya,
Everything works correctly, good job. Here's my review [FrontEnd Review](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2193#pullrequestreview-2037139321)

**Review by aaryaneil — APPROVED** [review]
Tested the PR and left a review on the [frontend](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2193#pullrequestreview-2039221428)

**Comment by KaushikMreddy** [comment]
Hi, 

I ran the backend step by step as mentioned, everything ran but no data is being populated in the tools table.
image

Data fetched from the API call seems to be empty
image

**Review by nathcarnevalli — COMMENTED** [review]
No data is being shown to me. I logged in with different accounts.

Screenshot 2024-05-14 105941

**Comment by ZhangPeizhou** [comment]
Reviewed in the frontend.

**Review by DaveB4r — APPROVED** [review]
Everything is perfect!
Great Job!
Building Management Dashboard OK!
bmdashboardlogin
Building and Inventory Management Dashboard OK!
bmdashboard
Tools OK!
bmdashboardtoolsfilter
Sort Tools OK!
bmdashboard-sort-runs
Purchase Record OK!
purchase

**Comment by one-community** [comment]
Thank you all, merging!

**Pull request merged**